### PR TITLE
core: refactor NetworkRequest time properties

### PR DESCRIPTION
**Pull request by connorjclark**
Broken up into multiple PRs:

- #14567

_____________

Implementing refactor as described here: https://github.com/GoogleChrome/lighthouse/pull/14227#discussion_r943005276

# new time properties / renames

```
Renamed:

startTime            -> networkRequestTime
endTime              -> networkEndTime
responseReceivedTime -> responseHeadersEndTime

Added:

rendererStartTime
rendererEndTime 
```

`rendererEndTime` is not actually something we have yet (only exists in trace), so at the moment it's always equal to `networkEndTime`. It is included to fill out our model of the network, and in some places the former is used instead of the latter even though it doesn't make a difference yet. At the very least, it better documents intent.

excerpt from network request constructor:

```js
/** When the renderer process initially discovers a network request, in milliseconds. */
this.rendererStartTime = -1;
/**
 * When the network service is about to handle a request, ie. just before going to the
 * HTTP cache or going to the network for DNS/connection setup, in milliseconds.
 */
this.networkRequestTime = -1;
/** When the last byte of the response headers is received, in milliseconds. */
this.responseHeadersEndTime = -1;
/** When the last byte of the response body is received, in milliseconds. */
this.networkEndTime = -1;
/**
 * Network response is done, this marks the time when the main thread is first available
 * to actually utilize the network resource, in milliseconds.
 *
 * TODO: we don't have this on the CDP, but it exists in trace as ResourceFinish.ts
 * So currently this is equivalent to networkEndTime.
 */
this.rendererEndTime = -1;
```

Tons of subtle changes to determine which time property should be used, thanks to @paulirish for going through all of them one-by-one with me.

I had to do an annoying refactor of critical request chain, in an attempt to avoid making a breaking change to its audit values.  It's a weird gatherer that saves the network request on it, then walks it in the audit...but needs to walk both the "gatherer chain" and the "audit detail chain"... which now has drifted in shape. The simple fix was to copy the chain traversal code and treat the two separately.

# s -> ms

This PR also changes our network record time values to be in ms instead of seconds. This should not impact any audit values - `network-requests` was already converted to ms.

However `critical-request-chain` timings and startTime/endTime in lantern `node`s remain as seconds. They'll be adjusted in a followup.

(Same deal in #14567)

### CODE REVIEW

File: core/test/audits/critical-request-chains-test.js [hunk 1]
```diff
@@ -121,14 +121,17 @@ describe('Performance: critical-request-chains audit', () => {
...
       expect(output.details.longestChain.transferSize).toEqual(2468);
       expect(output).toHaveProperty('score', 0);
     });
   });
 
   it('calculates the correct chain result for passing example (no 2.)', () => {
-    const artifacts = mockArtifacts(PASSING_CHAIN_RECORDS_2);
+    const artifacts = mockArtifacts(PASSING_CHAIN_RECORDS_2, {
+      // Precision issues from converting between ms and s create slight differences during roundtrip.
+      skipVerification: true,
```
Comment (connorjclark) [inline core/test/audits/critical-request-chains-test.js:133]: This was unfortunate.

image

**Review by brendankenny — COMMENTED** [review]
I'll wait for any changes the tests need to settle before digging in too closely, but overall looks great and I love the additional precision we'll have to describe network requests.

- Can we pick a different name than `internalNetworkRequestTime`? It's a key timestamp for request timing, so it's weird to put `internal` on the front of that. Makes it seem like you shouldn't be looking at it.
- Relatedly: for the `network-requests` hidden audit, that time is really important for post-analysis. e.g. a low-priority image could have a very quick `mainThreadStartTime` but a much later `internalNetworkRequestTime` due to being queued after a bunch of other requests. I guess we just need to include both in the output?

Also, doesn't it seem suspicious that `yarn test-lantern` is all 0% changes? If the network timings have all changed I would expect at least some movement.

File: core/test/audits/critical-request-chains-test.js (hunk 1 above)
Comment (brendankenny) [reply core/test/audits/critical-request-chains-test.js:133]: we should probably put an epsilon on those comparisons, but for a quick fix, could also just pick a different timestamp :)

**Comment by connorjclark** [comment]
Tests should be good now.

> Also, doesn't it seem suspicious that yarn test-lantern is all 0% changes? If the network timings have all changed I would expect at least some movement.

I forgot to mention - this PR should have zero impact on lantern. I figured it'd be good to split that out entirely.

The behavior remains exactly the same because `network-node` just uses `internalNetworkRequestTime`, which is the same as the previous `startTime`. There possibly might have been a change due to precision differences (like seen in that one test...) but I guess we don't deal with any input that would trigger that.

**Comment by connorjclark** [comment]
> Relatedly: for the network-requests hidden audit, that time is really important for post-analysis. e.g. a low-priority image could have a very quick mainThreadStartTime but a much later internalNetworkRequestTime due to being queued after a bunch of other requests. I guess we just need to include both in the output?

We could reflect all the time properties, and if it is important to keep `startTime` and `endTime` around we can have them be aliases for `internalNetworkRequestTime`\* and `networkEndTime` (or just drop  `internalNetworkRequestTime` and `networkEndTime`)

\* rename pending :)

File: core/audits/byte-efficiency/offscreen-images.js [hunk 2]
```diff
@@ -151,7 +151,7 @@ class OffscreenImages extends ByteEfficiencyAudit {
     return images.filter(image => {
       if (image.wastedBytes < IGNORE_THRESHOLD_IN_BYTES) return false;
       if (image.wastedPercent < IGNORE_THRESHOLD_IN_PERCENT) return false;
-      return image.requestStartTime < interactiveTimestamp / 1e6 - IGNORE_THRESHOLD_IN_MS / 1000;
+      return image.mainThreadStartTime < interactiveTimestamp / 1e6 - IGNORE_THRESHOLD_IN_MS / 1000;
```
Comment (adamraine) [inline core/audits/byte-efficiency/offscreen-images.js:154]: Does this expression need to change with the s -> ms conversion?

File: core/gather/driver/network-monitor.js [hunk 3]
```diff
@@ -218,9 +218,9 @@ class NetworkMonitor extends NetworkMonitorEventEmitter {
       if (request.protocol === 'ws' || request.protocol === 'wss') return;
 
       // convert the network timestamp to ms
```
Comment (adamraine) [inline core/gather/driver/network-monitor.js:220]: nit: rm comment

File: core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js [hunk 4]
```diff
@@ -155,7 +155,7 @@ class TagsBlockingFirstPaint extends FRGatherer {
    */
   static async findBlockingTags(driver, networkRecords) {
     const firstRequestEndTime = networkRecords.reduce(
-      (min, record) => Math.min(min, record.endTime),
+      (min, record) => Math.min(min, record.networkEndTime),
```
Comment (adamraine) [inline core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js:158]: Would it make sense to use `mainThreadEndTime` here?

File: types/artifacts.d.ts [hunk 5]
```diff
@@ -546,8 +546,10 @@ declare module Artifacts {
   }
 
   interface TagBlockingFirstPaint {
-    startTime: number;
-    endTime: number;
+    // ms
+    mainThreadStartTime: number;
+    // ms
```
Comment (adamraine) [inline types/artifacts.d.ts:551]: nit: jsdoc comments?

File: core/lib/network-request.js [hunk 6]
```diff
@@ -124,11 +124,25 @@ class NetworkRequest {
...
-    this.endTime = -1;
-    /** @type {number} */
-    this.responseReceivedTime = -1;
+    /** ms. When the main thread initially discovers a network request. */
+    this.mainThreadStartTime = -1;
+    /**
+     * ms. When the network service is about to handle a request, ie. just before going to the
+     * HTTP cache or going to the network for DNS/connection setup.
+     */
+    this.networkRequestTime = -1;
```
Comment (brendankenny) [inline core/lib/network-request.js:133]: for symmetry's sake, maybe this should be `networkStartTime`?

With `networkEndTime` it makes a nice counterpart to `mainThreadStartTime`/`mainThreadEndTime`, making it more obvious which properties mark start/end of durations without requiring much thought.

File: core/lib/network-request.js [hunk 7]
```diff
@@ -124,11 +124,25 @@ class NetworkRequest {
...
+    /** ms. When the first byte of the response headers is received. */
+    this.responseHeadersReceivedTime = -1;
```
Comment (brendankenny) [inline core/lib/network-request.js:135]: bikeshed: I think I prefer the old `responseReceivedTime` name (with clarification of how it's defined in a comment).

I can see the benefits of being clearer internally, but when using an instance externally, this property name makes me have to think a lot to parse it, and it kind of implies there are other response-received times, even though there's just the one :)

File: core/lib/network-request.js [hunk 8]
```diff
@@ -124,11 +124,25 @@ class NetworkRequest {
...
+    /** ms. When the last byte of the response body is received. */
+    this.networkEndTime = -1;
+    /**
+     * ms. Network response is done, this marks the time when the main thread is first available
+     * to actualy utilize the network resource.
```
Comment (brendankenny) [inline core/lib/network-request.js:140]: "when the network resource is first available for the main thread to utilize"?

File: core/lib/network-request.js [hunk 9]
```diff
@@ -319,18 +338,18 @@ class NetworkRequest {
 
   /**
    * @param {LH.Crdp.Network.Response} response
-   * @param {number} timestamp
+   * @param {number} ms
```
Comment (brendankenny) [inline core/lib/network-request.js:341]: All the other timestamps are doing the s -> ms conversion when set on their respective timing property, so for consistency it seems like it makes sense to do the same in `_onResponse`?

File: core/lib/network-request.js [hunk 10]
```diff
@@ -363,26 +382,29 @@ class NetworkRequest {
     if (timing.requestTime === 0 || timing.receiveHeadersEnd === -1) return;
     // Take startTime and responseReceivedTime from timing data for better accuracy.
     // Timing's requestTime is a baseline in seconds, rest of the numbers there are ticks in millis.
-    // TODO: This skips the "queuing time" before the netstack has taken over ... is this a mistake?
-    this.startTime = timing.requestTime;
-    const headersReceivedTime = timing.requestTime + timing.receiveHeadersEnd / 1000;
-    if (!this.responseReceivedTime || this.responseReceivedTime < 0) {
-      this.responseReceivedTime = headersReceivedTime;
+    this.networkRequestTime = timing.requestTime * 1000;
+    const headersReceivedTime = this.networkRequestTime + timing.receiveHeadersEnd;
```
Comment (brendankenny) [inline core/lib/network-request.js:386]: wait, is it actually when the last header byte is received?

File: core/test/lib/network-request-test.js [hunk 11]
```diff
@@ -111,10 +111,10 @@ describe('NetworkRequest', () => {
   describe('update fetch stats for Lightrider', () => {
```
Comment (brendankenny) [inline core/test/lib/network-request-test.js:111]: Add tests for the new timestamps?

**Review by brendankenny — COMMENTED** [review]
Really excited about the improved precision these extra timestamps bring. I disagree with my earlier comment that we should try to preserve the `network-requests` hidden audit properties; I think the new properties and names make it so much clearer it outweighs any breakages for the few of us looking at the audit.

In retrospect it would have been easier to review if we'd broken this up into
- rename `startTime`/`endTime` and add `mainThreadStartTime`/`mainThreadEndTime` with no other changes

and then some order of:
- switch to milliseconds in `network-request`
- switch audits to using mainThreadStart/End where appropriate
- CRC changes (though those are mostly self contained)

it's hard to track all of those simultaneously and feel confident we're getting it right in every file, and our network timing test coverage isn't always the best since we also need to avoid smoke flakes, but hopefully the power of combined reviewership will win out here.

File: core/audits/network-requests.js [hunk 12]
```diff
@@ -64,8 +64,8 @@ class NetworkRequests extends Audit {
       return {
         url: UrlUtils.elideDataURI(record.url),
         protocol: record.protocol,
-        startTime: timeToMs(record.startTime),
-        endTime: timeToMs(record.endTime),
+        startTime: normalizeTime(record.mainThreadStartTime),
+        endTime: normalizeTime(record.mainThreadEndTime),
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/audits/network-requests.js:68]: Do you at least want the old values here (aliased)?

File: core/audits/redirects.js [hunk 13]
```diff
@@ -125,9 +125,10 @@ class Redirects extends Audit {
       }
 
       const lanternTimingDeltaMs = redirectedTiming.startTime - initialTiming.startTime;
-      const observedTimingDeltaS = redirectedRequest.startTime - initialRequest.startTime;
+      const observedTimingDelta =
+        redirectedRequest.mainThreadStartTime - initialRequest.mainThreadStartTime;
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/audits/redirects.js:129]: the redirect's `mainThreadStartTime` is the same as requestWillBeSent timestamp. The name made sense for the original request but I suppose it's badly named for the redirected request, given the main thread is not involved in handling 3xx. Do you have a suggestion here, or maybe just a comment should be added on this line?

File: core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js (hunk 4 above)
Comment (connorjclark) [reply core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js:158]: Hmm... probably! I suppose we would be missing out on JS execution time here for scripts, which would be a big oversight.

Comment (connorjclark) [reply core/gather/gatherers/dobetterweb/tags-blocking-first-paint.js:158]: note: need to change down below too

File: core/lib/network-request.js [hunk 14]
```diff
@@ -124,11 +124,25 @@ class NetworkRequest {
     this.parsedURL = /** @type {ParsedURL} */ ({scheme: ''});
     this.documentURL = '';
 
-    this.startTime = -1;
-    /** @type {number} */
-    this.endTime = -1;
-    /** @type {number} */
-    this.responseReceivedTime = -1;
+    /** ms. When the main thread initially discovers a network request. */
+    this.mainThreadStartTime = -1;
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/lib/network-request.js:128]: 🤔 

any good alternative to "start"? `rendererInitiatedTime` / `rendererInitialTime` / `renderRequestedTime` / `rendererDiscoverTime` ?

File: core/lib/network-request.js (hunk 6 above)
Comment (connorjclark) [reply core/lib/network-request.js:133]: Am OK with this, @paulirish ?

File: core/lib/network-request.js (hunk 7 above)
Comment (connorjclark) [reply core/lib/network-request.js:135]: (i'd prefer `headersReceivedTime` to be more terse)

do you think we'd ever want to introduce more granular network timings? if so, we should be specific here

File: core/lib/network-request.js (hunk 8 above)
Comment (connorjclark) [reply core/lib/network-request.js:140]: this timing _includes_ and blocking long tasks that are happening on the page, even if (for example) a CSS background image is now available for the next layout/paint.

File: core/lib/network-request.js (hunk 10 above)
Comment (connorjclark) [reply core/lib/network-request.js:386]: ⚠️ ‼️ ⚠️ ‼️ ⚠️ ‼️ ⚠️ ‼️ 

there's no "start" equivalent on the timing object... It seems in practice it wouldn't make a difference, surely all sites flush out all the headers upfront 

does the lack of `receiveHeadersStart` imply that blink doesn't "stream" headers? if a preload header comes on the main document...then the server blocks before doing more headers then the body....would Chrome delay that preload?

File: core/audits/network-requests.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (paulirish) [reply core/audits/network-requests.js:68]: I like these sugggestions.

File: core/audits/redirects.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (paulirish) [reply core/audits/redirects.js:129]: In a followup refactor I'm interested in having redirects belong to the single request. (Instead of this weird cloned multi-request thing). I don't totally grok this discussion here, but it feels like the semantics are currently awkward.

File: core/lib/network-request.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (paulirish) [reply core/lib/network-request.js:128]: @connorjclark why replace 'start' ? seems ok. 😁 

@brendankenny whelp... we got the 'renderer process' and the 'main thread'. And yes other processes have their own 'main thread'... but.. we mostly never need to distinguish the browser/worker/etc main threads so... i'm think its aiight to treat 'mainthread' as a shorthand for 'rendererprocessmainthread'....
that said.... i guess hypothetically in some browser refactor we could see the very first timestamp of a network request creation starting on a diff renderer thread? so... this name could be future-proofing? 
yeah i'm cool with 'renderer... since its the process name and 'network' is the other process name.  

SUUUUUUUUURE lets do it.

File: core/lib/network-request.js (hunk 6 above)
Comment (paulirish) [reply core/lib/network-request.js:133]: Hmmmmmmmmmmm. I acknowledge that 'start' is more consistent.. but...........

semantically its not a "start"

image

the 'requestTime' is the line between 'queueing' and 'stalled'. but you can see that 'request sent' can be quite a while after it.  so... yah.

and also.. of all the times in https://nettimings.surge.sh/Network-Timings.svg "requestTime" is like the only one that has just 1 name everywhere. Good thing too since it's definition is so weird.

File: core/lib/network-request.js (hunk 7 above)
Comment (paulirish) [reply core/lib/network-request.js:135]: maybe we can drop "headers" if we sneak "start" into the name somewhere.

* `responseStartTime` i like. i guess we could just call out the 'headers received' technicality in a comment as the distinction isn't really important.
* `responseHeadersReceivedTime` does feel a bit wordy and overkill at this point.
* `headersReceivedTime` well... less wordy but it gets me thinking about "but what about request headers??"  so the thinking overhead is still there.
* `responseReceivedTime` (the old one) i dont like cuz it can easily be interpreted as 'end of response received'

File: core/lib/network-request.js (hunk 8 above)
Comment (paulirish) [reply core/lib/network-request.js:140]: I prefer the current PR's wording and agree with connor's point that it's a more main-thread-availability thing than a resource-availability thing.

File: core/lib/network-request.js (hunk 10 above)
Comment (paulirish) [reply core/lib/network-request.js:386]: > wait, is it actually when the last header byte is received?

yes

I've poked at this with the netlog, but i've never had a testcase where the response headers get split over two packets. 
But.. if i remember right even the //net stack doesnt really call out headersStart as a thing.

File: core/audits/network-requests.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply core/audits/network-requests.js]: > Do you at least want the old values here (aliased)?

I think it's fine, let's just make a clean break and remove the old ones

File: core/audits/redirects.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply core/audits/redirects.js]: > but it feels like the semantics are currently awkward.

yeah, that's the main issue, created timestamps that don't really fit the naming of a regular request because it's not a real request. Followup refactor sounds good.

File: core/lib/network-request.js (hunk 7 above)
Comment (brendankenny) [reply core/lib/network-request.js]: `responseHeadersEndTime` is another option if we want to be both specific, consistent with start/end, and precise, though it's unfortunate that usually what we want is `responseBodyStartTime`.

`responseStartTime` does seem like it would trip up a lot on the fact that the server has already responded :/

File: core/lib/network-request.js (hunk 8 above)
Comment (brendankenny) [reply core/lib/network-request.js]: Ah, ok, I just thought it was a typo but I see what you mean now. Maybe it would be clearer if the context was more explicit about the main thread being unblocked (since the resource is also "available", which is what tripped me up)?

`Network response is done; this marks the time when the network resource is available in the renderer and the main thread is unblocked and able to actually utilize it` or some variation on that?

File: core/audits/network-requests.js [hunk 15]
```diff
@@ -64,8 +64,10 @@ class NetworkRequests extends Audit {
       return {
         url: UrlUtils.elideDataURI(record.url),
         protocol: record.protocol,
-        startTime: timeToMs(record.startTime),
-        endTime: timeToMs(record.endTime),
+        rendererEndTime: normalizeTime(record.rendererEndTime),
+        networkEndTime: normalizeTime(record.networkEndTime),
+        networkRequestTime: normalizeTime(record.networkRequestTime),
+        rendererStartTime: normalizeTime(record.rendererStartTime),
```
Comment (brendankenny) [inline core/audits/network-requests.js:70]: what's going on with this order? :)

File: core/audits/network-requests.js [hunk 16]
```diff
@@ -87,7 +89,10 @@ class NetworkRequests extends Audit {
     const headings = [
       {key: 'url', itemType: 'url', text: 'URL'},
       {key: 'protocol', itemType: 'text', text: 'Protocol'},
-      {key: 'startTime', itemType: 'ms', granularity: 1, text: 'Start Time'},
+      {key: 'networkEndTime', itemType: 'ms', granularity: 1, text: 'Network End Time'},
+      {key: 'networkRequestTime', itemType: 'ms', granularity: 1, text: 'Network Request Time'},
+      {key: 'rendererEndTime', itemType: 'ms', granularity: 1, text: 'Renderer End Time'},
+      {key: 'rendererStartTime', itemType: 'ms', granularity: 1, text: 'Renderer Start Time'},
       {key: 'endTime', itemType: 'ms', granularity: 1, text: 'End Time'},
```
Comment (brendankenny) [inline core/audits/network-requests.js:96]: No more `endTime`

File: core/audits/network-requests.js (hunk 15 above)
Comment (connorjclark) [reply core/audits/network-requests.js:70]: alpha sort, oof. fixed.

File: core/test/network-records-to-devtools-log.js [hunk 17]
```diff
@@ -61,7 +61,8 @@ function getRequestWillBeSentEvent(networkRecord, index) {
         initialPriority: networkRecord.priority || 'Low',
         isLinkPreload: networkRecord.isLinkPreload,
       },
-      timestamp: networkRecord.redirectResponseTimestamp || networkRecord.startTime || 0,
+      timestamp:
+        networkRecord.redirectResponseTimestamp || networkRecord.mainThreadStartTime / 1000 || 0,
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/test/network-records-to-devtools-log.js]: Sorry, I'm not sure how to address this. Where would the values be set in this function?

File: core/test/lib/network-request-test.js (hunk 11 above)
Comment (connorjclark) [reply core/test/lib/network-request-test.js:111]: The test for `_updateTimingsForLightrider` or something else? There aren't really any tests here for startTime, endTime, etc. and I'm not sure what it'd be other than `check _onX() multiplies timestamp by 1000` or whatever.

Comment (connorjclark) [reply core/test/lib/network-request-test.js:111]: oh, doh, I guess I should just add the other fields here in getRequest()

File: core/lib/network-recorder.js [hunk 18]
```diff
@@ -248,7 +248,8 @@ class NetworkRecorder extends RequestEventEmitter {
 
     let candidates = recordsByURL.get(initiatorURL) || [];
     // The initiator must come before the initiated request.
-    candidates = candidates.filter(cand => cand.responseReceivedTime <= record.startTime);
+    candidates = candidates
+      .filter(cand => cand.responseHeadersEndTime <= record.networkRequestTime);
```
Comment (paulirish) [inline core/lib/network-recorder.js:252]: i think this could be rendererStartTime instead of networkRequestTime.
as we have to go back to renderer to make new requests. 
redirect would be the only case where that wouldnt happen but 1) thats cuz we artificially separate them into reqs and 2) i don't think we'll hit this situation with redirects.

File: core/test/lib/dependency-graph/simulator/simulator-test.js [hunk 19]
```diff
@@ -202,10 +202,10 @@ describe('DependencyGraph/Simulator', () => {
     });
 
     it('should not reuse connections', () => {
-      const nodeA = new NetworkNode(request({startTime: 0, endTime: 1}));
-      const nodeB = new NetworkNode(request({startTime: 2, endTime: 3}));
-      const nodeC = new NetworkNode(request({startTime: 2, endTime: 5}));
-      const nodeD = new NetworkNode(request({startTime: 2, endTime: 7}));
+      const nodeA = new NetworkNode(request({networkRequestTime: 0, networkEndTime: 1}));
```
Comment (paulirish) [inline core/test/lib/dependency-graph/simulator/simulator-test.js:205]: i noticed there's a bunch of other `startTime` and `endTime`'s in this file.. and while some belong to simulator nodes.. i think many should get renamed too..

File: core/test/lib/network-request-test.js (hunk 11 above)
Comment (connorjclark) [reply core/test/lib/network-request-test.js:111]: Added - but I had to make a change to `getResponseReceivedEvent` to fix roundtrip record->log errors. Don't understand why creating a timings was behind a flag,  but now it isn't and it fixed the roundtripping problem... seem ok?

File: core/test/lib/dependency-graph/simulator/simulator-test.js (hunk 19 above)
Comment (connorjclark) [reply core/test/lib/dependency-graph/simulator/simulator-test.js:205]: I'll rename these in the followup lantern change.

Comment (connorjclark) [reply core/test/lib/dependency-graph/simulator/simulator-test.js:205]: hold up how the hell are these other tests passing?!

like `should simulate cached network graphs`, the start/end time on nodeA/B just don't seem to matter (delete them, still passes), even the timing assert still passes 😕

File: core/test/network-records-to-devtools-log.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (brendankenny) [reply core/test/network-records-to-devtools-log.js]: basically if either `rendererStartTime` or `networkRequestTime` is set (but not both), the one not set should probably default to value of the other, not 0.

File: core/test/lib/dependency-graph/simulator/simulator-test.js (hunk 19 above)
Comment (connorjclark) [reply core/test/lib/dependency-graph/simulator/simulator-test.js:205]: it's b/c for that test the timing data comes from an estimate based on disk cache, real timing data is ignored. 
image

File: core/lib/dependency-graph/network-node.js [hunk 20]
```diff
@@ -25,14 +25,15 @@ class NetworkNode extends BaseNode {
    * @return {number}
    */
   get startTime() {
-    return this._record.startTime * 1000 * 1000;
+    // TODO change to mainThreadStartTime
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/lib/dependency-graph/network-node.js]: #14500

File: core/test/lib/dependency-graph/simulator/simulator-test.js (hunk 19 above)
Comment (connorjclark) [reply core/test/lib/dependency-graph/simulator/simulator-test.js]: I didn't like how simulator naming got more complicated because of network timing internals, so I kept the start/end time naming in these test files and just modified the `request()` helper function to translate back to the network domain

**Comment by paulirish** [comment]
I was thinking about the rename aspect of this PR separate from the behavior changes.. So I went through the changes and applied just the renames....

That diff: https://github.com/GoogleChrome/lighthouse/compare/renamenetwork

And then diffing this PR here against my rename-only branch: https://github.com/GoogleChrome/lighthouse/compare/renamenetwork..network-request-timing-overhaul

Maybe this diff is informative... I gotta run but i was also thinking.. maybe it'd be good for this branch to call things `ambiguousStartTime` so that it doesn't start to decide who uses what.
